### PR TITLE
Add `exception` field in invokeresult

### PR DESCRIPTION
**Pull request by Tommo-L**
Fix https://github.com/neo-project/neo-modules/issues/334

### CODE REVIEW

**Review by shargon — DISMISSED** [review]

**Comment by superboyiii** [comment]
Test failed. Exception is not displayed.
```
            Console.WriteLine("**********RpcClient*************");
            RpcClient rpcClient = new RpcClient("http://localhost:12332");
            Console.WriteLine();

            /*invokescript*/
            Console.WriteLine("********************************");
            Console.WriteLine("invokescript:");
            byte[] script = "10c00c066465706c6f790c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b52".HexToBytes();
            RpcInvokeResult rpcInvokeResult = rpcClient.InvokeScript(script);
            var result = rpcInvokeResult.ToJson().ToString();
            Console.WriteLine("inovke result: " + result);
            Console.WriteLine("********************************");

            /*invoke balanceof*/
            Console.WriteLine("invoke balanceof:");
            RpcStack rpcStack = new RpcStack()
            {
                Type = "Hash160",
                Value = "0xf642a0401fdd56a03114639a98b7963eb587a30a"
            };
            Signer[] signers = new[]
                            {
                    new Signer
                    {
                        Account = UInt160.Parse("0x9f2dd7df630980dce111d3b31ebe35d3cfae58cd"),
                        Scopes = WitnessScope.CalledByEntry
                    }
                };
            rpcInvokeResult = rpcClient.InvokeFunction("0x254b9decd76080ef368e7a6b0a065938dfbc31cf", "balanceOf", new RpcStack[] { rpcStack }, signers);
            string balanceOf = rpcInvokeResult.ToJson().ToString();
            Console.WriteLine("inovke balanceOf: " + balanceOf);
            Console.WriteLine("********************************");

            /*invoke destory(exeception case)*/
            rpcStack = new RpcStack()
            {
                Type = "Hash160",
                Value = "0x9f2dd7df630980dce111d3b31ebe35d3cfae58cd"
            };
            signers = new[]
                            {
                    new Signer
                    {
                        Account = UInt160.Parse("0xf642a0401fdd56a03114639a98b7963eb587a30a"),
                        Scopes = WitnessScope.CalledByEntry
                    }
                };
            rpcInvokeResult = rpcClient.InvokeFunction("0x254b9decd76080ef368e7a6b0a065938dfbc31cf", "destory", new RpcStack[] { rpcStack }, signers);
            string destory = rpcInvokeResult.ToJson().ToString();
            Console.WriteLine("inovke destory: " + destory);
            Console.ReadLine();
```
image
but got this:
image
```
Unhandled exception. Neo.Network.RPC.RpcException: Failed execution for '0c14cd58aecfd335be1eb3d311e1dc800963dfd72d9f11c00c07646573746f72790c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b52'
   at Neo.Network.RPC.RpcClient.Send(RpcRequest request) in E:\PR-test\modules-PR#364\neo-modules\src\RpcClient\RpcClient.cs:line 87
   at Neo.Network.RPC.RpcClient.RpcSend(String method, JObject[] paraArgs) in E:\PR-test\modules-PR#364\neo-modules\src\RpcClient\RpcClient.cs:line 100
   at Neo.Network.RPC.RpcClient.InvokeFunction(String scriptHash, String operation, RpcStack[] stacks, Signer[] signer) in E:\PR-test\modules-PR#364\neo-modules\src\RpcClient\RpcClient.cs:line 345
   at RpcClientTest.Program.Main(String[] args) in E:\PR-test\modules-PR#364\RpcClientTest\Program.cs:line 63

E:\PR-test\modules-PR#364\RpcClientTest\bin\Debug\netcoreapp3.1\RpcClientTest.exe (process 33680) exited with code 0.
```

**Comment by devhawk** [comment]
Can this please wait until after #335 is merged?

**Comment by superboyiii** [comment]
> Can this please wait until after #335 is merged?

Fine.

**Comment by superboyiii** [comment]
image

File: src/RpcClient/Models/RpcInvokeResult.cs
Comment (shargon) [inline src/RpcClient/Models/RpcInvokeResult.cs:29]: string.IsNullOrEmpty?

Comment (Tommo-L) [reply src/RpcClient/Models/RpcInvokeResult.cs]: Done

**Review by shargon — APPROVED** [review]

**Comment by superboyiii** [comment]
image

**Comment by Qiao-Jin** [comment]
This looks like Exception thrown by neo-modules rather than by contract.

**Comment by superboyiii** [comment]
@Tommo-L 
Still...
```
Unhandled exception. System.AggregateException: One or more errors occurred. (Failed execution for '10c00c0764657374726f790c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b52')
 ---> Neo.Network.RPC.RpcException: Failed execution for '10c00c0764657374726f790c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b52'
   at Neo.Network.RPC.RpcClient.SendAsync(RpcRequest request) in E:\PR-test\modules-PR#364\neo-modules\src\RpcClient\RpcClient.cs:line 80
   at Neo.Network.RPC.RpcClient.RpcSendAsync(String method, JObject[] paraArgs) in E:\PR-test\modules-PR#364\neo-modules\src\RpcClient\RpcClient.cs:line 93
   at Neo.Network.RPC.RpcClient.InvokeFunctionAsync(String scriptHash, String operation, RpcStack[] stacks, Signer[] signer) in E:\PR-test\modules-PR#364\neo-modules\src\RpcClient\RpcClient.cs:line 352
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at RpcClientTest.Program.Main(String[] args) in E:\PR-test\modules-PR#364\RpcClientTest\Program.cs:line 56
```

File: src/RpcServer/RpcServer.Wallet.cs
Comment (shargon) [inline src/RpcServer/RpcServer.Wallet.cs:189]: Exception?

Comment (Tommo-L) [reply src/RpcServer/RpcServer.Wallet.cs]: Done

File: src/RpcClient/Models/RpcInvokeResult.cs
Comment (superboyiii) [inline src/RpcClient/Models/RpcInvokeResult.cs:51]: Here should be modified since if json contains `"exception":null,` **NullReferenceException** will come out like this:
```
{{"script":"DBQKo4e1Ppa3mJpjFDGgVt0fQKBC9hHADAliYWxhbmNlT2YMFM8xvN84WQYKa3qONu\u002BAYNfsnUslQWJ9W1I=","state":"HALT","gasconsumed":"3840690","exception":null,"stack":[{"type":"Integer","value":"2000000000000000"}],"tx":"00c30c1478b29a3a00000000000ea31200000000006f1c200001cd58aecfd335be1eb3d311e1dc800963dfd72d9f01003e0c140aa387b53e96b7989a631431a056dd1f40a042f611c00c0962616c616e63654f660c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b5201420c4023e054a3d253bb0c9b5320922536f6dfc9fba3647d9f70012247abcb49d87b9082118f3417d06a8b198b341d1af0750f5b4bba451e6397a259cecf1cb02c1ac6290c2102fdf27a2eb8d9b069d61ae5a64dba13b89637c542d8c74c5d6bab493fe638d68f0b4195440d78"}}
```
image

Comment (Tommo-L) [reply src/RpcClient/Models/RpcInvokeResult.cs]: Fixed

**Comment by superboyiii** [comment]
I found that if `Signer[]` contains more than 1 elements, InvokeFunctionAsync will throw **NullReferenceException** in ToJson because of `json["type"]` is null.
```
            /*invoke destory(normal)*/
            Console.WriteLine("Destroy(normal case)");
            RpcStack[] rpcStack1 = new RpcStack[] { };
            signers = new[]
                {
                    new Signer
                    {
                        Account = UInt160.Parse("0xf642a0401fdd56a03114639a98b7963eb587a30a"),
                        Scopes = WitnessScope.CalledByEntry
                    },
                    new Signer
                    {
                        Account = UInt160.Parse("0x9f2dd7df630980dce111d3b31ebe35d3cfae58cd"),
                        Scopes = WitnessScope.CalledByEntry
                    }
                };
            rpcInvokeResult = rpcClient.InvokeFunctionAsync("0x254b9decd76080ef368e7a6b0a065938dfbc31cf", "destroy", rpcStack1, signers);
            string destroy = rpcInvokeResult.Result.ToJson().ToString();
            Console.WriteLine("inovke destroy: " + destroy);
            Console.WriteLine("********************************");
```
image

**Comment by superboyiii** [comment]
Oh, I found it's fixed in https://github.com/neo-project/neo-modules/pull/371, that's fine.

**Review by superboyiii — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Review by superboyiii — APPROVED** [review]

**Comment by superboyiii** [comment]
Test Pass.
```
**********RpcClient*************

********************************
Open Wallet
open wallet: True
********************************
Invoke Balanceof
inovke balanceOf: {"script":"DBQKo4e1Ppa3mJpjFDGgVt0fQKBC9hHADAliYWxhbmNlT2YMFM8xvN84WQYKa3qONu\u002BAYNfsnUslQWJ9W1I=","state":"HALT","gasconsumed":"3840690","stack":[{"type":"Integer","value":"2000000000000000"}],"tx":"00e65dab31b29a3a00000000000ea31200000000006f1c200001cd58aecfd335be1eb3d311e1dc800963dfd72d9f01003e0c140aa387b53e96b7989a631431a056dd1f40a042f611c00c0962616c616e63654f660c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b5201420c40ad6526320ad75df804cab696cc05813f8adea322b9d9d3ef54c50809c5610d594f51da80f06e873414d4ca17db2c5e6c3f8d8ff0d49267825b7207d0f5366d76290c2102fdf27a2eb8d9b069d61ae5a64dba13b89637c542d8c74c5d6bab493fe638d68f0b4195440d78"}
********************************
Destroy(normal- > single signer)
inovke destroy: {"script":"EMAMB2Rlc3Ryb3kMFM8xvN84WQYKa3qONu\u002BAYNfsnUslQWJ9W1I=","state":"HALT","gasconsumed":"2302320","stack":[{"type":"Any"}],"tx":"00688b205b70212300000000004e451200000000006f1c2000010aa387b53e96b7989a631431a056dd1f40a042f601002610c00c0764657374726f790c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b5201420c401fa3e2c142266099549d22503550679bb96bc8e76c8b9f6f86b6f7309a549bc71ce6807321faf120312881eb5b9014eaab9321db1bfab217214cba908ae2b131290c2103cea3e2f81f1af93522d29ede4f24a1f0bedead7293391d33880f5475d97878c90b4195440d78"}
********************************
Destroy(normal -> multi signers)
inovke destroy: {"script":"EMAMB2Rlc3Ryb3kMFM8xvN84WQYKa3qONu\u002BAYNfsnUslQWJ9W1I=","state":"HALT","gasconsumed":"2302320","stack":[{"type":"Any"}],"tx":"00c07aa6467021230000000000e4842300000000006f1c2000020aa387b53e96b7989a631431a056dd1f40a042f601cd58aecfd335be1eb3d311e1dc800963dfd72d9f01002610c00c0764657374726f790c14cf31bcdf3859060a6b7a8e36ef8060d7ec9d4b2541627d5b5202420c401a2f170e6d7aadd450969fae425fe456ef432090848fdec560090f9e218e48663b9199fcbb051333f01f6f2e820e92f43f7954e743c65903bc51391c21b9b4d7290c2103cea3e2f81f1af93522d29ede4f24a1f0bedead7293391d33880f5475d97878c90b4195440d78420c408b11d73a6b75ec9285464f079ff82c6e8cd320cfc581c0d02b645cbfc294acb3f2d4d8940e4f98035554c17e4b7be235de9ba868a0e7d8be56768195ef799f18290c2102fdf27a2eb8d9b069d61ae5a64dba13b89637c542d8c74c5d6bab493fe638d68f0b4195440d78"}
********************************
Destroy(exeception -> signer is not owner)
inovke destroy: {"script":"EMAMB2Rlc3Ryb3kMFM8xvN84WQYKa3qONu\u002BAYNfsnUslQWJ9W1I=","state":"FAULT","gasconsumed":"1324470","exception":"An unhandled exception was thrown. No authorization.","stack":[]}
********************************
```

**Review by shargon — APPROVED** [review]

**Pull request merged**
